### ISSUE TITLE
amp-ad: Inaccessible ad content

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#29565]
## WCAG Level
Level AA

## Priority
Medium

## Type
Colour, Navigation

## Pages/Components affected
* [amp-ad](https://amp.dev/documentation/components/amp-ad/?format=websites)

## Description
While not directly controlled by the AMP project itself, note that the content embedded into pages using `<amp-ad>` can, depending on the provider, introduce accessibility issues. For instance, on the [playground example](https://playground.amp.dev/?url=https%3A%2F%2Fpreview.amp.dev%2Fdocumentation%2Fexamples%2Fcomponents%2Famp-ad&format=websites&_gl=1*1wy27k5*_ga*YW1wLWdDQmZ0cDgwZl96MWxoRWR1NEFlaXc.), the embedded ad content (from `<amp-ad type="A9">`) has insufficient text color contrast and a lack of visible focus indication.

An important design consideration is to make sure there is enough contrast between the colour of text and the colour of its background. Good colour contrast makes content more readable for people who find it difficult to see, or who have conditions that affect their ability to perceive colour. The `<amp-ad type="A9">` links have insufficient contrast, both by default and when hovered with the mouse.

|Example|Details|
|-------|-------|
|Hovered link ![Hovered link](https://user-images.githubusercontent.com/54933393/88788034-b295a680-d18c-11ea-95eb-a6da25fb798f.png)| Foreground: `#FF9900` /Background: `#FFFFFF` / contrast ratio: 2.1:1 |
|Regular (non-hovered) links ![Non-hovered links](https://user-images.githubusercontent.com/54933393/88788068-be816880-d18c-11ea-9c56-1a0112b10809.png)| Foreground: `#4a90e2` /Background: `#FFFFFF` / contrast ratio: 3.3:1 |

Keyboard users rely on the browser's focus outline to keep track of where they are on web pages. Most browsers use a faint dotted outline to indicate the focus to the user, though some browsers like Chrome or Safari use an enhanced focus outline, consisting of a coloured border around the selected link. It is also possible for developers to provide custom focus indications within the site's cascading style sheets (CSS). Currently, the links in the `<amp-ad type="A9">` content actively suppress any focus indication with CSS.

```
div#amzn-native-ad-0 * div:focus, div#amzn-native-ad-0 * a:focus, div#amzn-native-ad-0 * span:focus {
    outline: 0;
}
```

## User impact
Without sufficient contrast, content can be difficult to perceive and therefore difficult to read. In extreme cases, content with low contrast may not be visible at all to people with colour perception impairments. This may result in users not being able to see/read the presented ad content.

Without a visible focus indicator it can be difficult or impossible for keyboard users that aren't using additional assistive technology such as a screen reader to know which control (whether it be a link, button, or something else) currently has focus and therefore what action will take place if they press the `Enter`, `Space`, or other keys.

## Required solution
Make sure the foreground and background colours used for text follow these rules:

* Standard text (including images of text, links, button labels, and other controls) has a contrast ratio of at least 4.5:1 against its background.
* Large text (18pt/24px or 14pt/18.5px bold) has a contrast ratio of at least 3:1.

Incidental text that is not part of an interactive element, is pure decoration, or is part of a logo or picture, is exempt from these rules.

In the case of visible focus indication, the simplest solution is to remove the current CSS that suppresses the `outline`.

This content is likely provided by the third-party ad networks. If at all possible, these providers should be encouraged/told that their current CSS has accessibility issues relating to color contrast and visible focus indication.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied for text contrast:

1. Check the contrast of all text using a tool like the [TPG: Colour Contrast Analyser](https://developer.paciellogroup.com/resources/contrastanalyser/).
2. Check that the contrast level is at least 4.5:1 for small text (14pt/18.5px) or 3:1 for large text (18pt/24px or 14pt/18.5px bold)

Use these steps to confirm that the solution has been correctly applied for visible focus indication:
1. Navigate to a focusable component on the page using the keyboard (`Tab`/`Shift`+`Tab`).
2. Check that there is a visible focus indicator.

## Related standards
* [WCAG 2.1 Success Criterion 1.4.3 Contrast (Minimum) (Level AA)](https://www.w3.org/TR/WCAG21/#contrast-minimum)
* [WCAG 2.1 Success Criterion 2.4.7 Focus Visible (Level AA)](https://www.w3.org/TR/WCAG21/#focus-visible)

## More information
* [TPG: Colour Contrast Analyser](https://developer.paciellogroup.com/resources/contrastanalyser/)
* [WebAIM: Color blindness](https://webaim.org/articles/visual/colorblind)
* [W3C - Failure of Success Criterion 2.4.7 due to styling element outlines and borders in a way that removes or renders non-visible the visual focus indicator](https://www.w3.org/WAI/WCAG21/Techniques/failures/F78.html)

## Test data
Test date: July 2020

**Comment by caroqliu** [comment]
FYI @ampproject/wg-ads

**Comment by powerivq** [comment]
@dthukral-amazon in case you have some idea, or know who in Amazon owns this extension.

**Comment by jeffkaufman** [comment]
This seems like a tricky situation, where you're not going to get ads with accessible levels of contrast unless the ad networks make that policy. Looking over Google's policies, for example, I don't see anything that requires a minimum level of contrast (but I could have missed it) and I'm not aware of other networks that require this today.

I could imagine the amp project making this a requirement for inclusion of ad extensions, but my guess is that most and networks would then go back to loading arbitrary content in an iframe?

Another angle: what does AMP think of someone using amp-youtube (or similar) to embed a video which has text that is too low contrast?
